internal/providers/vault/provider.go

```diff
@@ -43,6 +46,20 @@ func (p *Provider) GetName() string {
 // hash returned by the Vault.
 const VaultDefaultField = "value"
 
+// VaultBackendKVv2 is the Vault backend KV version 2.
+const VaultBackendKVv2 = "kv-v2"
+
+// parseVaultBackend returns the (optional) backend of the backend from the id.
+func parseVaultBackend(id string) (string, string) {
+	tokens := strings.SplitN(id, ":", 2)
+	switch len(tokens) {
+	case 1:
+		return "", id
```

Comment:
> Using kv: prefix would break existing behavior. I advice against it.

Can you elaborate on this (I'm a bit newish on the Vault API specifics)? How is a generic secret fetch different from KVv1?

> Keep in mind that 99% of the code is equal. It's just the unwrapping of the KV v2 response that is different. I don't think extensibility is improved by factoring out into kv and kv-v2 variants.

I'd disagree here to an extent. Even if we ignore the extensibility part which I for sure should have elaborated on more, there's a bit of a problem. Whenever multiple logic paths are embedded this way in a codebase, the mental overhead increases for the maintainer to know how they intertwine in the flow as they are parsing the code and the test matrix grows by orders of magnitude for each branch in the code. I think some of this pattern is generally described by [single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) and more often than not it has held accurate in my experience.

Reply:
> Can you elaborate on this (I'm a bit newish on the Vault API specifics)? How is a generic secret fetch different from KVv1?

Vault looks like a file system where you read from / write to, with contents made up of JSON. The "old" implementation the Vault provider for Conjur expects the secret value as JSON field in the returned object. Which works for various secret engines supported by Vault, hence the generic secret fetch. A notable exception is the KVv2 secret engine, due to its additional wrapping of the secret value in another object within the response object. That's why it needs to be handled differently.

> ... single-responsibility principle ...
I will look into a better code design, will get back to you on this.

Reply:
I opted for an alternative design, in which the distinction between various Vault secret engines/backend doesn't appear as much as it did in my previous solution. I redesigned to have more expressive field navigation. It uses `path/to/secert#navigate.to.field` instead of prefixing with `kv-v2:`. Hence the field name has become field navigation, as if traversing through an object structure. This is both compatible to the current behavior as well as extends naturally to the KV v2 case.